cowbird/handlers/impl/geoserver.py

```diff
@@ -141,38 +258,110 @@ def get_instance():
+                    publish_shapefile.si(workspace_name, shapefile_name))
+        res.delay()
+
     def on_created(self, filename):
         """
-        Call when a new file is found.
+        Called when a new file is found.
 
         :param filename: Relative filename of a new file
         """
-        if filename.endswith(".shp"):
+        # TODO: ajouter un case pour un workspace (folder)?
+        #  On a normalement un workspace par user et ils sont déjà créés et ajoutés à Geoserver lors de la création d'un
+        #  user. Est-ce qu'on supporte un cas où un folder serait créé manuellement? Est-ce que c'est automatiquement un
+        #  workspace à ajouter à Geoserver et à Magpie?
+        if filename.endswith(tuple(SHAPEFILE_ALL_EXTENSIONS)):
             workspace_name, shapefile_name = self._get_shapefile_info(filename)
+
             LOGGER.info("Starting Geoserver publishing process for [%s]", filename)
-            res = chain(validate_shapefile.si(workspace_name, shapefile_name),
-                        publish_shapefile.si(workspace_name, shapefile_name))
-            res.delay()
+            Geoserver.publish_shapefile_task_chain(workspace_name, shapefile_name)
+
+            self._update_magpie_layer_permissions(workspace_name, shapefile_name)
+
+    @staticmethod
+    def remove_shapefile_task(workspace_name, shapefile_name):
+        # type: (str, str) -> None
+        """
+        Applies the celery task required to remove a shapefile from Geoserver.
+        """
+        remove_shapefile.delay(workspace_name, shapefile_name)
 
     def on_deleted(self, filename):
         """
-        Call when a file is deleted.
+        Called when a file is deleted.
 
         :param filename: Relative filename of the removed file
         """
-        if filename.endswith(".shp"):
+        # TODO: voir pour le case workspace (folder)
+        if filename.endswith(tuple(SHAPEFILE_ALL_EXTENSIONS)):
             workspace_name, shapefile_name = self._get_shapefile_info(filename)
-            remove_shapefile.delay(workspace_name, shapefile_name)
+            Geoserver.remove_shapefile_task(workspace_name, shapefile_name)
+
+            # Remove all the remaining shapefile related files
+            for file in self.get_shapefile_list(workspace_name, shapefile_name):
+                if os.path.exists(file):
+                    os.remove(file)
+
+            # Remove the corresponding Magpie resource
+            magpie_handler = HandlerFactory().get_handler("Magpie")
+            layer_res_id = magpie_handler.get_geoserver_resource_id(workspace_name, shapefile_name,
+                                                                    create_if_missing=False)
+            if layer_res_id:
+                magpie_handler.delete_resource(layer_res_id)
 
     def on_modified(self, filename):
         # type: (str) -> None
         """
-        Call when a file is updated.
+        Called when a file is updated.
```

Comment:
Should a directory modification just update its corresponding Magpie permissions? I suppose we could use a `Recursive` scope by default for workspaces, since on the file system if blocks the access to the children files.

If the file is modified, the Magpie permissions should still be updated and we assume if the folder does not have read permission, there should be a corresponding `Deny-recursive` on the Magpie workspace resource, which would block the access to the layer resource too.
So maybe we should create a resource and corresponding permissions for the `folder` case too after all?

Also, if having no permissions defined on a layer means no access to the file, having no permission for a workspace should also mean no access to the folder and therefore its content on the file system. But on Magpie if there are no permissions defined for the workspace, the layer resource could be accessible, so I think the workspace should always either have `allow-recursive` or `deny-recursive`? This way the access would actually correspond to the access on the filesystem?

I will have to check for the rename case, which I haven't taken care of so far.

Reply:
In most cases, I would expect that only the top-level user-workspace-dir should have read-recursive for the user. 
Any other resource under it (dir/file or workspace/layer) would probably never need to be generated in Magpie. 
If the effective permission resolves to allow from the top-level user-workspace-dir, there is no advantage to specify more granular resource/permissions. If anything, it would just slow down the resolution as it must rewind the resource/permission tree each time for each combination.

So, unless the directory is `chmod` explicitly to adjust `r/x` for reading/browsing it, the resource/permission should probably not be generated. For example, if `chmod u+r,u+x` is applied, and Magpie already resolves the corresponding workspace dir as allow, don't generate anything. Inversely, if it resolves to deny, then yes, generate the resource/permission to allow access, since that would mean something higher up causes the deny resolution. The same idea can apply for access removal with `chmod u-r,u-x`, whereas already denied doesn't need any resource/permission update, while something allowed would now require an explicit deny definition.

Reply:
> Does this event trigger on a rename (either file or dir), in which case the resource/permissions must be moved over, or is it handled as a sequence of on_delete/on_create for before/after names?

I finally investigated the rename case. I tested it with the monitoring and I see it triggers :
- an update on the parent directory
- both a delete and create event on the file

So I think it will work as it is, at least the files/layers. The associated geoserver shapefile and magpie resources will be deleted, and new ones will be created with the new name.

But I think we should not support the case for a folder rename, since it will probably break everything anyway if we decide to either rename the workspace directory (which is named by the user name) or the datastore directory (which is by default using the `shapefile_datastore` name). I think it should just not be done.

Reply:
Fair enough. It goes along another comment where we established that manual changes to those directories is not recommended. Manual changes is somewhat a "do at your own risks", and only a platform admin/manager can do them. If they break workspaces/permission-sync, that will be their problem and Cowbird shouldn't need to cover for them. If anything, resync could probably be resolved by manually triggering the sync when [DAC-442](https://crim-ca.atlassian.net/browse/DAC-442) gets worked on.